base_layer/key_manager/src/birthday.rs

```diff
@@ -0,0 +1,140 @@
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+use tari_common::configuration::Network;
+use tari_core::blocks::genesis_block::get_genesis_block;
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Birthday {
+    zero_point_time: u64,
+    birthday: u16,
+    version: u8,
+}
+
+impl Birthday {
+    pub fn default() -> Self {
+        let network = Network::Dibbler;
+        Self::new(network)
+    }
+
+    pub fn new(network: Network) -> Self {
+        let current_time = Self::current_time_in_seconds();
+        Self::new_from_current_time(network, current_time)
+    }
+
+    fn new_from_current_time(network: Network, current_time: u64) -> Self {
+        const SECONDS_PER_DAY: u64 = 24 * 60 * 60;
+        const PERIOD_LENGTH: u64 = u16::MAX as u64 + 1; // 2^16
+
+        let mut zero_point_time = Self::get_network_genesis_time(network);
+
+        let days = (current_time - zero_point_time) / SECONDS_PER_DAY;
+        let birthday = u16::try_from(days % PERIOD_LENGTH).unwrap();
+        let version = u8::try_from(days / PERIOD_LENGTH).unwrap();
+
+        zero_point_time += PERIOD_LENGTH * u64::from(version);
+
+        Self {
+            birthday,
+            version,
+            zero_point_time,
+        }
+    }
+
+    pub fn birthday(&self) -> u16 {
+        self.birthday
+    }
+
+    pub fn version(&self) -> u8 {
+        self.version
+    }
+
+    pub fn zero_point_time(&self) -> u64 {
+        self.zero_point_time
+    }
+
+    pub fn current_time_in_seconds() -> u64 {
+        u64::try_from(chrono::Utc::now().timestamp()).unwrap()
+    }
+
+    pub fn get_network_genesis_time(network: Network) -> u64 {
+        get_genesis_block(network).block().header.timestamp.as_u64()
+    }
```

Comment:
That is definitely a possibility. However, I don't feel confident that the birthday genesis time might depend on the API use. Since we are primarily interested in using `Birthday` to set up a timestamp to scan the blockchain, it might be good to have a consistent genesis time for birthday. I am possibly missing something in your commentary.

Moreover, following [issue](https://github.com/tari-project/tari/issues/4176), it is referred that one should have an hard-coded birthday genesis time, and it is suggested to use the timestamp of the genesis network. I also really want to remove the dependency on the `Network`, as well. One could just hard-code a recent date (like now) and set it for the future use of the `Birthday` type.

Reply:
The idea behind hardcoding all parameters is to ensure that all implementations and installations agree on them unambiguously. Then you don't need to worry about differences in defaults or other computed parameters; everyone is working from the same playbook.